### ISSUE TITLE
Open / close the port often for ethernet comms

### ISSUE THREAD

**Issue by chrisflesher** [jfjlaros/arduino-simple-rpc#5]
The `Interface` class is currently designed to work with serial ports that remain open for long periods of time. This approach is causing the Arduino ethernet server to get stuck in the communications `while` loop when waiting for the port to close. It would be nice for ethernet connections to open / close frequently to allow the Arduino to do other things when using ethernet or WiFi.

**Describe alternatives you've considered**

1. Add an option to keep message definitions instead of deleting them when the port closes. This is the least invasive change but makes the python API somewhat complex:
```
with Interface('socket://192.168.1.50:10000', wait=0, reuse_methods=True) as interface:
    print(interface.ping(1))
```

2. Same as option 1 but change `Interface` defaults (e.g. `wait=0` and `reuse_methods=True` by default). This could break existing functionality but would result in a cleaner python API.
```
with Interface('socket://192.168.1.50:10000') as interface:
    print(interface.ping(1))
```

3. Modify `Interface` class to open / close ports frequently (including serial ports). This would be a heavy modification of the `Interface` class, would remove the open / close functions and internally use `with serial_for_url(device)` to open / close the port for each read / write operation. This would cause several breaking changes but would result in the simplest python API.
```
interface = Interface('socket://192.168.1.50:10000')
print(interface.ping(1))
```

**Describe the solution you'd like**
@jfjlaros I haven't selected a solution, maybe will prototype options 1 and 3 and see what you think after that.

**Comment by jfjlaros** [comment]
What about adding a `reopen()` method to the interface class?

**Comment by chrisflesher** [comment]
Hmm... I guess you could call `open` (to get methods) then `close` (to delete methods but you could save a copy somewhere) then `reopen` (this would copy over the old methods without getting them again). Is this what you're suggesting?

This would work, the python API would look like this:
```
from simple_rpc import Interface

interface = Interface('socket://192.168.1.50:10000')
print(interface.ping(1))
interface.close()
interface.reopen()
print(interface.ping(1))
interface.close()
```

That would work, the API for ethernet connections would be somewhat verbose but its fine for what I'm trying to do for now.

**Comment by jfjlaros** [comment]
Something like that. Or we could remove the whole removal of methods altogether.

**Comment by jfjlaros** [comment]
Perhaps something like this: `reopen` only opens the serial connection, `open` calls `reopen`, does the wait, fetches the methods, etc. and `close` does not throw away the methods anymore.

**Comment by jfjlaros** [comment]
By the way, I have enabled [discussions](https://github.com/jfjlaros/arduino-simple-rpc/discussions) for this repository.

**Comment by chrisflesher** [comment]
Ok, I'll put up a different PR to prototype that idea. I haven't used the discussions feature before, it looks kind of like Slack?

**Comment by jfjlaros** [comment]
No idea, I never used it either.

**Comment by jfjlaros** [comment]
We could give it a try perhaps, see [this](https://github.com/jfjlaros/arduino-simple-rpc/discussions/9) discussion.

**Comment by jfjlaros** [comment]
Fixed in #13.

**Comment by chrisflesher** [comment]
Thanks for looking at this!
